invoices/modification_interceptor.go

```diff
@@ -0,0 +1,197 @@
+		callback: &safeCallback{},
+	}
+}
+
+// Intercept generates a new intercept session for the given invoice. The call
+// blocks until the client has responded to the request or an error occurs. The
+// response callback is only called if a session was created in the first place,
+// which is only the case if a client is registered.
+func (s *HtlcModificationInterceptor) Intercept(clientRequest HtlcModifyRequest,
+	responseCallback func(HtlcModifyResponse)) error {
+
+	// If there is no client callback set we will not handle the invoice
+	// further.
+	if !s.callback.IsConnected() {
+		log.Debugf("Not intercepting invoice with circuit key %v, no "+
+			"intercept client connected",
+			clientRequest.ExitHtlcCircuitKey)
+
+		return nil
+	}
+
+	// We'll block until the client has responded to the request or an error
+	// occurs.
+	var (
+		responseChan = make(chan *HtlcModifyResponse, 1)
+		errChan      = make(chan error, 1)
+	)
+
+	// The callback function will block at the client's discretion. We will
+	// therefore execute it in a separate goroutine. We put the callback on
+	// the stack by passing it into the goroutine through a parameter. That
+	// means it will be executed in any case, even when the client has
+	// disconnected in the meantime.
+	s.wg.Add(1)
+	go func() {
+		defer s.wg.Done()
+
+		log.Debug("Waiting for client response from invoice HTLC "+
+			"interceptor session with circuit key %v",
+			clientRequest.ExitHtlcCircuitKey)
+
+		// By this point, we've already checked that the client callback
+		// is set. However, if the client disconnected since that check
+		// then Exec will return an error.
+		result, err := s.callback.Exec(clientRequest)
+		if err != nil {
+			err = fmt.Errorf("client callback failed: %w", err)
+			log.Error(err)
+
+			_ = fn.SendOrQuit(errChan, err, s.quit)
+
+			return
+		}
+
+		_ = fn.SendOrQuit(responseChan, result, s.quit)
+	}()
+
+	// Wait for the client to respond or an error to occur.
+	select {
+	case response := <-responseChan:
+		log.Debugf("Received invoice HTLC interceptor response: %v",
+			response)
+
+		responseCallback(*response)
+
+		return nil
+
+	case err := <-errChan:
+		log.Errorf("Error from invoice HTLC interceptor session: %v",
+			err)
+
+		return err
+
+	case <-s.quit:
+		return ErrInterceptorClientDisconnected
+	}
+}
+
+// RegisterInterceptor sets the client callback function that will be called
+// when an invoice is intercepted. If a callback is already set, an error is
+// returned. The returned function must be used to reset the callback to nil
+// once the client is done or disconnects.
+func (s *HtlcModificationInterceptor) RegisterInterceptor(
+	callback HtlcModifyCallback) (func(), error) {
+
+	return s.callback.Set(callback)
+}
+
+// Start starts the service.
+func (s *HtlcModificationInterceptor) Start() error {
+	return nil
+}
+
+// Stop stops the service.
+func (s *HtlcModificationInterceptor) Stop() error {
+	close(s.quit)
+
+	s.wg.Wait()
```

Comment:
one thing that stands out to me here is that this will wait till `s.callback.Exec(clientRequest)` returns which is not something that is currently controlled from this side via either a channel or a context. Wondering if we can somehow pass something through to Exec to ensure that it quits when we want to call Stop here

Reply:
awesome